File: playwright/03-intercept-seo-save.js
/**
 * Injects decrypted Chrome cookies into Playwright, opens page properties dialog
 * for the Home page, intercepts the save request to find the exact API format,
 * then sets SEO fields for all 4 pages.
 *
 *   node playwright/03-intercept-seo-save.js
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const COOKIES_JSON = '/tmp/zoho_cookies_clean.json';
const BASE = 'https://sitebuilder-60059075182.zohositescontent.in';
const SITE_ID = '413198000000002010';

const PAGES = [
  {
    name: 'Home',
    id: '413198000000004653',
    url_slug: 'home',
    seo_title: 'Butler Button | Personal Travel Expert - 100% Human, 150+ Countries',
    seo_desc: '100% human travel experts. 150+ countries. AI plans it, a human handles it, you enjoy it. On-demand concierge from $25. No membership required.',
  },
  {
    name: 'Trip Planning',
    id: '413198000000047002',
    url_slug: 'trip-planning',
    seo_title: 'AI Trip Planning with Human Expert Review | Butler Button',
    seo_desc: 'AI screens 200,000+ options. A human travel expert reviews it. Delivered in 24 hours. $25/country. 150+ countries. No membership required.',
  },
  {
    name: 'Concierge',
    id: '413198000000002013',
    url_slug: 'concierge',
    seo_title: 'On-Demand Human Travel Concierge - Not a Chatbot | Butler Button',
    seo_desc: 'A real human concierge who pre-reads your itinerary before you depart. 24/7 coverage. 97% of disruptions resolved under 60 min. Starting $25/day.',
  },
  {
    name: 'Travel Advisor',
    id: '413198000000047014',
    url_slug: 'travel-advisor',
    seo_title: 'Private Travel Advisor Under Your Brand | Butler Button by VELTM',
    seo_desc: 'White-label travel advisor service for professionals. Butler Button delivers 5-star concierge under your brand. Your clients get a personal travel expert.',
  },
];

function buildCookies(rawCookies) {
  const cookies = [];
  for (const [name, value] of Object.entries(rawCookies)) {
    if (!value || typeof value !== 'string') continue;
    // Deduplicate _zxor (same name twice in source)
    if (cookies.find(c => c.name === name && c.value === value)) continue;
    cookies.push({
      name,
      value,
      domain: '.sitebuilder-60059075182.zohositescontent.in',
      path: '/',
      httpOnly: false,
      secure: true,
    });
  }
  return cookies;
}

async function interceptSaveRequest(page) {
  return new Promise(resolve => {
    let captured = null;
    page.on('request', req => {
      const url = req.url();
      const method = req.method();
      if ((method === 'PUT' || method === 'POST') && url.includes('/pages')) {
        const body = req.postData();
        captured = { url, method, body, headers: req.headers() };
        console.log(`\n  INTERCEPTED: ${method} ${url}`);
        console.log(`  Body: ${body ? body.substring(0, 500) : '(empty)'}`);
        console.log(`  CSRF header: ${req.headers()['x-zcsrf-token'] || 'MISSING'}`);
        resolve(captured);
      }
    });
    // Also resolve after 15s if no request fires
    setTimeout(() => resolve(captured), 15000);
  });
}

async function setPageSEO(context, pageData, knownSaveFormat) {
  const page = await context.newPage();

  // Enable request interception
  const savePromise = interceptSaveRequest(page);

  const pagesUrl = `${BASE}/zcms/${SITE_ID}/pages`;
  await page.goto(pagesUrl, { waitUntil: 'networkidle', timeout: 30000 });
  await page.waitForTimeout(3000);

  console.log(`  Current URL: ${page.url()}`);

  // Take screenshot for debug
  await page.screenshot({ path: path.join(__dirname, `debug-pages-${pageData.id}.png`) });

  // Dump page structure to find the page row
  const pageContent = await page.evaluate(() => {
    // Look for any element containing the page name or resource_id
    const rows = document.querySelectorAll('[data-resource-id], [data-page-id], [data-id], .sites-pageitem, .sites-listitem');
    return Array.from(rows).slice(0, 20).map(r => ({
      tag: r.tagName,
      classes: r.className,
      dataResourceId: r.getAttribute('data-resource-id'),
      dataPageId: r.getAttribute('data-page-id'),
      dataId: r.getAttribute('data-id'),
      text: r.innerText ? r.innerText.substring(0, 50) : '',
    }));
  });
  console.log('  Page items found:', JSON.stringify(pageContent.slice(0, 5)));

  // Try to find and click the settings for this page
  // Zoho Sites typically has a gear/settings icon per page row
  const settingsTriggers = await page.evaluate((pageId) => {
    // Look for elements with data attributes matching page id
    const byId = document.querySelectorAll(`[data-resource-id="${pageId}"], [data-page-id="${pageId}"], [data-id="${pageId}"]`);
    return Array.from(byId).map(el => ({
      tag: el.tagName,
      classes: el.className,
      text: el.innerText ? el.innerText.substring(0, 50) : '',
    }));
  }, pageData.id);
  console.log(`  Triggers for ${pageData.id}:`, settingsTriggers);

  // Try clicking settings icon - look for various patterns
  const clicked = await page.evaluate((pageId, pageName) => {
    // Pattern 1: find parent container with resource-id then find action button
    const containers = document.querySelectorAll(`[data-resource-id="${pageId}"]`);
    if (containers.length > 0) {
      const cont = containers[0].closest('[class*="item"], [class*="row"], li') || containers[0].parentElement;
      const btn = cont && cont.querySelector('[class*="setting"], [class*="gear"], [class*="more"], button');
      if (btn) { btn.click(); return `clicked btn in container: ${btn.className}`; }
    }
    // Pattern 2: find by page name text
    const allText = document.querySelectorAll('*');
    for (const el of allText) {
      if (el.children.length === 0 && el.innerText && el.innerText.trim() === pageName) {
        const row = el.closest('[class*="item"], [class*="row"], li') || el.parentElement;
        if (row) {
          const btn = row.querySelector('[class*="setting"], [class*="gear"], [class*="more"], [class*="opt"], button');
          if (btn) { btn.click(); return `clicked via name: ${btn.className}`; }
          // Try right-clicking the row
          row.dispatchEvent(new MouseEvent('contextmenu', { bubbles: true }));
          return `context menu on row: ${row.className}`;
        }
      }
    }
    return 'nothing clicked';
  }, pageData.id, pageData.name);
  console.log(`  Click result: ${clicked}`);

  await page.waitForTimeout(2000);

  // Screenshot after click
  await page.screenshot({ path: path.join(__dirname, `debug-afterclick-${pageData.id}.png`) });

  // Look for any dialog/panel that opened with SEO fields
  const dialogs = await page.evaluate(() => {
    const elements = document.querySelectorAll('[class*="dialog"], [class*="modal"], [class*="panel"], [class*="popup"], [class*="overlay"], [role="dialog"]');
    return Array.from(elements)
      .filter(el => el.offsetParent !== null) // visible only
      .map(el => ({
        tag: el.tagName,
        classes: el.className,
        id: el.id,
        html: el.innerHTML.substring(0, 500),
      }));
  });
  console.log(`  Visible dialogs: ${dialogs.length}`);
  dialogs.forEach(d => console.log(`    ${d.tag}.${d.classes.substring(0,50)} id=${d.id}`));

  // If no dialog, dump full page body
  if (dialogs.length === 0) {
    const bodyHTML = await page.evaluate(() => document.body.innerHTML.substring(0, 3000));
    fs.writeFileSync(path.join(__dirname, `debug-body-${pageData.id}.html`), bodyHTML);
    console.log(`  No dialogs -- body HTML saved`);
  }

  const captured = await savePromise;
  await page.close();
  return captured;
}

async function main() {
  const rawCookies = JSON.parse(fs.readFileSync(COOKIES_JSON, 'utf-8'));
  const cookies = buildCookies(rawCookies);
  console.log(`Loaded ${cookies.length} cookies`);

  const browser = await chromium.launch({ headless: false, slowMo: 50 });
  const context = await browser.newContext({
    viewport: { width: 1400, height: 900 },
  });
  await context.addCookies(cookies);

  // First: navigate to the editor and check auth
  const page = await context.newPage();
  await page.goto(`${BASE}/zcms/${SITE_ID}/pages`, { waitUntil: 'domcontentloaded', timeout: 30000 });
  await page.waitForTimeout(3000);
  const url = page.url();
  console.log(`Auth check URL: ${url}`);

  if (url.includes('accounts.zoho') || url.includes('login')) {
    console.error('Session expired! Cannot proceed without re-auth.');
    await browser.close();
    process.exit(1);
  }
  await page.screenshot({ path: path.join(__dirname, 'debug-auth-check.png') });
  console.log('Auth OK. Starting page properties interception...');
  await page.close();

  // Process Home page first to capture the save format
  console.log(`\n--- Processing Home page for format discovery ---`);
  const captured = await setPageSEO(context, PAGES[0], null);
  if (captured) {
    console.log('\n=== CAPTURED SAVE REQUEST ===');
    console.log(JSON.stringify(captured, null, 2));
    fs.writeFileSync(path.join(__dirname, 'debug-save-request.json'), JSON.stringify(captured, null, 2));
  }

  await browser.close();
}

main().catch(err => { console.error(err); process.exit(1); });

File: playwright/03-intercept-seo-save.py
"""
Sets native Zoho SEO title + description for all 4 pages via the
"Edit page info" UI dialog. Injects decrypted Chrome cookies so no
manual login is needed.

  python3 playwright/03-intercept-seo-save.py
"""

import json, sys, time
from pathlib import Path
from playwright.sync_api import sync_playwright

COOKIES_JSON = '/tmp/zoho_cookies_clean.json'
BASE = 'https://sitebuilder-60059075182.zohositescontent.in'
SITE_ID = '413198000000002010'
SCRIPT_DIR = Path(__file__).parent

PAGES = [
    {
        'name': 'Home',
        'id': '413198000000004653',
        'seo_title': 'Butler Button | Personal Travel Expert - 100% Human, 150+ Countries',
        'seo_desc': '100% human travel experts. 150+ countries. AI plans it, a human handles it, you enjoy it. On-demand concierge from $25. No membership required.',
    },
    {
        'name': 'Trip Planning',
        'id': '413198000000047002',
        'seo_title': 'AI Trip Planning with Human Expert Review | Butler Button',
        'seo_desc': 'AI screens 200,000+ options. A human travel expert reviews it. Delivered in 24 hours. $25/country. 150+ countries. No membership required.',
    },
    {
        'name': 'Concierge',
        'id': '413198000000002013',
        'seo_title': 'On-Demand Human Travel Concierge - Not a Chatbot | Butler Button',
        'seo_desc': 'A real human concierge who pre-reads your itinerary before you depart. 24/7 coverage. 97% of disruptions resolved under 60 min. Starting $25/day.',
    },
    {
        'name': 'Travel Advisor',
        'id': '413198000000047014',
        'seo_title': 'Private Travel Advisor Under Your Brand | Butler Button by VELTM',
        'seo_desc': 'White-label travel advisor service for professionals. Butler Button delivers 5-star concierge under your brand. Your clients get a personal travel expert.',
    },
]


def build_cookies(raw: dict) -> list:
    seen = set()
    cookies = []
    for name, value in raw.items():
        if not value or not isinstance(value, str):
            continue
        key = (name, value)
        if key in seen:
            continue
        seen.add(key)
        cookies.append({
            'name': name,
            'value': value,
            'domain': '.sitebuilder-60059075182.zohositescontent.in',
            'path': '/',
            'httpOnly': False,
            'secure': True,
        })
    return cookies


def screenshot(page, label):
    try:
        page.screenshot(path=str(SCRIPT_DIR / f"{label}.png"), timeout=10000)
    except Exception as e:
        print(f"  [screenshot failed: {e}]")


def process_page(context, page_data, all_requests):
    print(f"\n--- {page_data['name']} ({page_data['id']}) ---")
    page = context.new_page()

    def capture(req):
        if req.method in ('PUT', 'POST') and '/pages' in req.url and 'api' in req.url:
            body = req.post_data or ''
            hdrs = dict(req.headers)
            entry = {
                'page': page_data['name'],
                'method': req.method,
                'url': req.url,
                'body': body,
                'csrf': hdrs.get('x-zcsrf-token', ''),
            }
            all_requests.append(entry)
            print(f"  *** CAPTURED {req.method} {req.url}")
            print(f"      body: {body[:400]}")
            print(f"      csrf: {hdrs.get('x-zcsrf-token','')[:80]}")

    page.on('request', capture)

    pages_url = f"{BASE}/zcms/{SITE_ID}/pages"
    page.goto(pages_url, wait_until='networkidle', timeout=30000)
    page.wait_for_timeout(2000)
    screenshot(page, f"s01-{page_data['id']}-list")

    # Click "Edit page info" for this specific page row.
    # The link text is exactly "Edit page info" and it's in the same row as the page name.
    clicked = page.evaluate("""(pageName) => {
        // Find the row containing this page name
        for (const el of document.querySelectorAll('a, span, td, div')) {
            if (el.children.length === 0 && (el.innerText || '').trim() === pageName) {
                // Walk up to the row
                let row = el.closest('tr, li, [class*="item"], [class*="row"]') || el.parentElement;
                if (!row) continue;
                // Find "Edit page info" link in this row
                for (const a of row.querySelectorAll('a, button, span')) {
                    if ((a.innerText || '').trim() === 'Edit page info') {
                        a.click();
                        return 'clicked Edit page info in row: ' + row.tagName + '.' + row.className.substring(0,40);
                    }
                }
            }
        }
        // Fallback: click any "Edit page info" link (first visible one matching the page)
        for (const a of document.querySelectorAll('a[href*="' + pageName.toLowerCase().replace(/ /g,'-') + '"], a')) {
            if ((a.innerText || '').trim() === 'Edit page info' && a.offsetParent) {
                a.click();
                return 'fallback Edit page info: ' + a.href;
            }
        }
        return 'not found';
    }""", page_data['name'])
    print(f"  Edit page info click: {clicked}")
    page.wait_for_timeout(3000)
    screenshot(page, f"s02-{page_data['id']}-dialog")

    # The dialog has tabs: General, SEO, Options. Click SEO tab.
    seo_tab = page.evaluate("""() => {
        // data-tab="seo" or li with text "SEO"
        for (const el of document.querySelectorAll('[data-tab="seo"], li, a, button')) {
            const t = (el.innerText || '').trim();
            if (t === 'SEO' && el.offsetParent) {
                el.click();
                return 'clicked SEO tab: ' + el.tagName + '.' + el.className.substring(0,40);
            }
        }
        return 'SEO tab not found';
    }""")
    print(f"  SEO tab: {seo_tab}")
    page.wait_for_timeout(1000)
    screenshot(page, f"s03-{page_data['id']}-seo-tab")

    # Dump dialog HTML for debug
    dialog_info = page.evaluate("""() => {
        const visible = [];
        for (const el of document.querySelectorAll('[class*="dialog"], [class*="modal"], [class*="panel"], [class*="popup"], [role="dialog"], [class*="pagedetails"], [class*="page-prop"]')) {
            if (el.offsetParent !== null) {
                visible.push({
                    tag: el.tagName,
                    classes: el.className.substring(0, 60),
                    id: el.id,
                    html: el.innerHTML.substring(0, 2000),
                });
            }
        }
        return visible;
    }""")
    if dialog_info:
        print(f"  Visible dialogs: {len(dialog_info)}")
        for d in dialog_info:
            print(f"    {d['tag']}#{d['id']}.{d['classes'][:40]}")
        # Save first dialog HTML for inspection
        (SCRIPT_DIR / f"dbg-dialog-{page_data['id']}.html").write_text(
            dialog_info[0]['html'] if dialog_info else ''
        )
    else:
        # No standard dialog -- dump all visible forms/inputs
        forms_info = page.evaluate("""() => {
            const inputs = document.querySelectorAll('input[type="text"], textarea');
            return Array.from(inputs).filter(el => el.offsetParent !== null).map(el => ({
                tag: el.tagName,
                name: el.name,
                id: el.id,
                dataBind: el.getAttribute('data-bind'),
                placeholder: el.placeholder,
                value: el.value.substring(0, 50),
            }));
        }""")
        print(f"  No dialogs. Visible inputs: {len(forms_info)}")
        for inp in forms_info[:10]:
            print(f"    {inp}")

    # Fill SEO title field
    filled = page.evaluate("""([title, desc]) => {
        const log = [];
        const titleSels = [
            '[data-bind="page_info.seo.title"]',
            '[data-bind*="seo.title"]',
            '#seo-c input[type="text"]',
            'input.form-control[type="text"]',
            'input.sites-inputtext',
        ];
        const descSels = [
            '[data-bind="page_info.seo.description"]',
            '[data-bind*="seo.description"]',
            '#seo-c textarea',
            'textarea.form-control',
            'textarea.sites-textarea',
        ];

        function setField(el, val) {
            // Use native input value setter to trigger React/Angular bindings
            const nativeSetter = Object.getOwnPropertyDescriptor(el.__proto__ || Object.getPrototypeOf(el), 'value');
            if (nativeSetter && nativeSetter.set) {
                nativeSetter.set.call(el, val);
            } else {
                el.value = val;
            }
            ['input', 'change', 'keyup', 'blur'].forEach(evt =>
                el.dispatchEvent(new Event(evt, {bubbles: true}))
            );
        }

        let titleEl = null;
        for (const sel of titleSels) {
            const el = document.querySelector(sel);
            if (el && el.offsetParent !== null) { titleEl = el; log.push('title:' + sel); break; }
        }
        let descEl = null;
        for (const sel of descSels) {
            const el = document.querySelector(sel);
            if (el && el.offsetParent !== null) { descEl = el; log.push('desc:' + sel); break; }
        }

        if (titleEl) setField(titleEl, title);
        else log.push('NO_TITLE');
        if (descEl) setField(descEl, desc);
        else log.push('NO_DESC');

        return {log, titleFound: !!titleEl, descFound: !!descEl,
                titleVal: titleEl ? titleEl.value : null,
                descVal: descEl ? descEl.value.substring(0, 80) : null};
    }""", [page_data['seo_title'], page_data['seo_desc']])
    print(f"  Fill: {filled}")
    page.wait_for_timeout(500)

    if not filled.get('titleFound') and not filled.get('descFound'):
        print(f"  WARN: no SEO fields found. Check dbg-dialog-{page_data['id']}.html")
        page.close()
        return False

    # Click Save button
    saved = page.evaluate("""() => {
        // Look for Save button by text in visible area
        for (const btn of document.querySelectorAll('button, [class*="btn"]')) {
            const t = (btn.innerText || '').trim();
            if ((t === 'Save' || t === 'Update' || t === 'Done') && btn.offsetParent) {
                btn.click();
                return 'Save clicked: ' + t + ' class=' + btn.className.substring(0,40);
            }
        }
        return 'no save button';
    }""")
    print(f"  Save: {saved}")
    page.wait_for_timeout(3000)
    screenshot(page, f"s04-{page_data['id']}-after-save")

    page.close()
    return True


def main():
    raw = json.loads(Path(COOKIES_JSON).read_text())
    cookies = build_cookies(raw)
    print(f"Loaded {len(cookies)} cookies")

    all_requests = []

    with sync_playwright() as pw:
        browser = pw.chromium.launch(headless=False, slow_mo=80)
        context = browser.new_context(viewport={'width': 1440, 'height': 900})
        context.add_cookies(cookies)

        # Auth check
        page = context.new_page()
        page.goto(f"{BASE}/zcms/{SITE_ID}/pages", wait_until='domcontentloaded', timeout=30000)
        page.wait_for_timeout(2000)
        url = page.url
        print(f"Auth: {url}")
        if 'accounts.zoho' in url or 'login' in url:
            print("Session expired.")
            browser.close()
            sys.exit(1)
        print("Auth OK")
        page.close()

        results = []
        for p in PAGES:
            ok = process_page(context, p, all_requests)
            results.append({'name': p['name'], 'ok': ok})

        browser.close()

    print("\n=== RESULTS ===")
    for r in results:
        print(f"  {'OK' if r['ok'] else 'FAIL'} {r['name']}")

    if all_requests:
        out = SCRIPT_DIR / 'debug-save-requests.json'
        out.write_text(json.dumps(all_requests, indent=2))
        print(f"\nCaptured {len(all_requests)} save request(s) -> {out}")
        print(json.dumps(all_requests[0], indent=2))


if __name__ == '__main__':
    main()
